lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+    }
+    return n;
+  }
+
+  constexpr std::int64_t ToInt64() const {
+    std::int64_t signExtended = ToUInt64();
+    if (bits < 64) {
+      signExtended |= -(signExtended >> (bits - 1)) << bits;
+    }
+    return signExtended;
+  }
+
+  // Ones'-complement (i.e., C's ~)
+  constexpr Integer NOT() const {
+    Integer result{nullptr};
+    for (int j{0}; j < parts; ++j) {
+      result.SetLEPart(j, ~LEPart(j));
+    }
+    return result;
+  }
+
+  // Two's-complement negation (-x = ~x + 1).
+  // An overflow flag accompanies the result, and will be true when the
+  // operand is the most negative signed number (MASKL(1)).
+  constexpr ValueWithOverflow Negate() const {
+    Integer result{nullptr};
+    Part carry{1};
+    for (int j{0}; j + 1 < parts; ++j) {
+      Part newCarry{LEPart(j) == 0 && carry};
+      result.SetLEPart(j, ~LEPart(j) + carry);
+      carry = newCarry;
+    }
+    Part top{LEPart(parts - 1)};
+    result.SetLEPart(parts - 1, ~top + carry);
+    bool overflow{top != 0 && result.LEPart(parts - 1) == top};
+    return {result, overflow};
+  }
+
+  constexpr ValueWithOverflow ABS() const {
+    if (IsNegative()) {
+      return Negate();
+    } else {
+      return {*this, false};
+    }
+  }
+
+  // Shifts the operand left when the count is positive, right when negative.
+  // Vacated bit positions are filled with zeroes.
+  constexpr Integer ISHFT(int count) const {
+    if (count < 0) {
+      return SHIFTR(-count);
+    } else {
+      return SHIFTL(count);
+    }
+  }
+
+  // Left shift with zero fill.
+  constexpr Integer SHIFTL(int count) const {
+    if (count <= 0) {
+      return *this;
+    } else {
+      Integer result{nullptr};
+      int shiftParts{count / partBits};
+      int bitShift{count - partBits * shiftParts};
+      int j{parts - 1};
+      if (bitShift == 0) {
+        for (; j >= shiftParts; --j) {
+          result.SetLEPart(j, LEPart(j - shiftParts));
+        }
+        for (; j >= 0; --j) {
+          result.LEPart(j) = 0;
+        }
+      } else {
+        for (; j > shiftParts; --j) {
+          result.SetLEPart(j,
+              ((LEPart(j - shiftParts) << bitShift) |
+                  (LEPart(j - shiftParts - 1) >> (partBits - bitShift))));
+        }
+        if (j == shiftParts) {
+          result.SetLEPart(j, LEPart(0) << bitShift);
+          --j;
+        }
+        for (; j >= 0; --j) {
+          result.LEPart(j) = 0;
+        }
+      }
+      return result;
+    }
+  }
+
+  // Circular shift of a field of least-significant bits.  The least-order
+  // "size" bits are shifted circularly in place by "count" positions;
+  // the shift is leftward if count is nonnegative, rightward otherwise.
+  // Higher-order bits are unchanged.
+  constexpr Integer ISHFTC(int count, int size = bits) const {
+    if (count == 0 || size <= 0) {
+      return *this;
+    }
+    if (size > bits) {
+      size = bits;
+    }
+    if ((count %= size) == 0) {
```

Comment:
I think it's clearer to separate the assignment from the condition, so:
```
count %= size;
if (count == 0) {
```

Reply:
Ok.